zkml/src/layers/pooling.rs

```diff
@@ -185,40 +190,43 @@ impl Pooling {
 
         diff_polys.push(beta_poly);
         vp.add_mle_list(diff_polys, E::ONE);
+
         lookup_parts
             .into_iter()
             .for_each(|(prod, coeff)| vp.add_mle_list(prod, coeff));
 
+        // We also batch in the smae poly proof for the output with the Zerocheck
+        let output_mle: ArcMultilinearExtension<E> = output.get_data().to_vec().into_mle().into();
+
+        vp.add_mle_list(
+            vec![output_mle.clone(), last_claim_beta],
+            challenge_combiner,
+        );
         #[allow(deprecated)]
         let (proof, sumcheck_state) = IOPProverState::<E>::prove_parallel(vp, prover.transcript);
 
-        // We need to prove that the output of this step is the input to following activation function
-        let output_mle = output.get_data().to_vec().into_mle();
-        let mut same_poly_prover = same_poly::Prover::<E>::new(output_mle.clone());
-
+        // Extract all claims about committed witness polys
         let zerocheck_point = &proof.point;
-        let output_zerocheck_eval = output_mle.evaluate(zerocheck_point);
+        let sumcheck_evals = sumcheck_state.get_mle_final_evaluations();
+        let kernel_size = info.poolinfo.kernel_size * info.poolinfo.kernel_size;
 
-        // Accumulate claims about the output polynomial in each of the protocols we ran together with the final claim from the previous proof.
-        let mut output_claims = Vec::<Claim<E>>::new();
-        let same_poly_ctx = same_poly::Context::<E>::new(last_claim.point.len());
-        same_poly_prover.add_claim(last_claim.clone())?;
-
-        let zerocheck_claim = Claim {
-            point: zerocheck_point.clone(),
-            eval: output_zerocheck_eval,
-        };
-        same_poly_prover.add_claim(zerocheck_claim.clone())?;
-        output_claims.push(zerocheck_claim);
-
-        // This is the proof for the output poly
-        let claim_acc_proof = same_poly_prover.prove(&same_poly_ctx, prover.transcript)?;
-
-        let output_claim = claim_acc_proof.extract_claim();
+        sumcheck_evals[..kernel_size]
+            .iter()
+            .enumerate()
+            .try_for_each(|(i, &eval)| {
+                let claim = Claim::<E>::new(zerocheck_point.clone(), eval);
+                prover
+                    .witness_prover
+                    .add_claim(info.poly_id * 100 + i, claim)
+            })?;
+        let output_eval = sumcheck_evals[kernel_size + 1];
 
         prover
             .witness_prover
-            .add_claim(info.poly_id, output_claim)
+            .add_claim(
+                info.poly_id * 100 + kernel_size,
```

Comment:
Well if we start having equations even to derive these IDs that doesn't look good :p 
Could you implement just a simple random ID drawing mechanism that we use during context where we assign the IDs to these polynomials and we can always refer to them or stg ?